### PR TITLE
--copy-ignored flag added to CLI 

### LINKED ISSUE #6226: copy-files flag does not respect ignore block in .babelrc
**Bug:** If in your build script, you have `--copy-files` passed, files that are ignored in .babelrc files will be copied over, instead of being ignored.

### Input Code
```js
babel src --out-dir lib --copy-files
```

### Babel Configuration (.babelrc, package.json, cli command)

```js
{
  "presets": [
    "es2015",
    "stage-1",
    "react"
  ],
  "ignore": [
    "**__mocks__/*.js"
  ]
}
```

### Expected Behavior
Babel should not copy files that are explicitly ignored in `.babelrc` configuration. In this case, babel should not transpile or copy Jest mocks.

### Current Behavior
Babel copies files that are ignored are explicitly `.babelrc`.
```sh
> npm run build 
> babel src --out-dir lib --copy-files
  src/api/__mocks__/index.js -> lib/api/__mocks__/index.js
  src/api/endpoints.js -> lib/api/endpoints.js
```
If you pass an `--ignore` flag (with the same blob) _after_ the `--copy-files` flag, it will behave as expected.
```sh
> npm run build;
> babel src --out-dir lib --copy-files --ignore **__mocks__/*.js
  src/api/endpoints.js -> lib/api/endpoints.js
  (output does not include ignored directories)
```

### Context
This issue comes up when you work with Jest manual mocks inside of your source directory. Jest warns that you have duplicate __mocks__ directories.

| software         | version(s)
| ---------------- | -------
| Babel            |  6.26.0
| node             | 6.9.2
| npm              | 5.4.0
| Operating System | OSX El Capitan

### PR DESCRIPTION
**Pull request by rajasekarm**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #6226 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      | Yes
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

### CODE REVIEW

**Comment by nicolo-ribaudo** [comment]
Reviews at https://github.com/babel/babel/pull/10831

File: packages/babel-cli/src/babel/dir.js [hunk 1]
```diff
@@ -84,8 +84,16 @@ export default async function({
...
-    if (!written && cliOptions.copyFiles) {
+    const relative = path.relative(base, src);
+    const isCompilableExtension = util.isCompilableExtension(
+      relative,
+      cliOptions.extensions,
+    );
+    if (
+      !written &&
+      ((!isCompilableExtension && cliOptions.copyFiles) ||
+        cliOptions.includeIgnored)
```
Comment (JLHwung) [inline packages/babel-cli/src/babel/dir.js]: I think when `copyFiles` is not enabled but `includeIgnored` is enabled, we should throw in `parseArgv`, otherwise we will have another flag that _could_ copy non compilable files.

File: packages/babel-cli/test/fixtures/babel/--copy-files with ignore with includeIgnored/options.json [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{
+  "args": [
```
Comment (JLHwung) [inline packages/babel-cli/test/fixtures/babel/--copy-files with ignore with includeIgnored/options.json]: Can you add a test for `--only`?
```
babel src --out-dir lib --copy-files --only src/main/* --include-ignored
```

File: packages/babel-cli/src/babel/dir.js (hunk 1 above)
Comment (existentialism) [reply packages/babel-cli/src/babel/dir.js]: 🚲🏠 warning:

Should we call it `--copy-ignored`? 😝

Comment (rajasekarm) [reply packages/babel-cli/src/babel/dir.js]: Changed to --copy-ignored

**Comment by rajasekarm** [comment]
I'm throwing error if `--copy-ignored` is used without `ignore` flag. Which will solve other corner cases.

File: packages/babel-cli/test/fixtures/babel/--copy-files with ignore with includeIgnored/options.json (hunk 2 above)
Comment (rajasekarm) [reply packages/babel-cli/test/fixtures/babel/--copy-files with ignore with includeIgnored/options.json]: Is this still needed?

File: packages/babel-cli/test/fixtures/babel/--copy-files --include-dotfiles with ignore/out-files/lib/foo/.foo.js [hunk 3]
```diff
@@ -1 +0,0 @@
-a;
```
Comment (JLHwung) [inline packages/babel-cli/test/fixtures/babel/--copy-files --include-dotfiles with ignore/out-files/lib/foo/.foo.js]: 🤔Why is `.foo.js` not copied?

Comment (rajasekarm) [reply packages/babel-cli/test/fixtures/babel/--copy-files --include-dotfiles with ignore/out-files/lib/foo/.foo.js]: we've --include-dotfile to Include dotfiles when compiling and copying non-compilable files.

https://github.com/rajasekarm/babel/blob/master/packages/babel-cli/src/babel/options.js#L148

Comment (JLHwung) [reply packages/babel-cli/test/fixtures/babel/--copy-files --include-dotfiles with ignore/out-files/lib/foo/.foo.js]: Yeah, it _was_ copied before this PR. But now .foo.js has been removed from the out files, which means it is _not_ copied.

Comment (rajasekarm) [reply packages/babel-cli/test/fixtures/babel/--copy-files --include-dotfiles with ignore/out-files/lib/foo/.foo.js]: I got it, I'll check.

Comment (rajasekarm) [reply packages/babel-cli/test/fixtures/babel/--copy-files --include-dotfiles with ignore/out-files/lib/foo/.foo.js]: I've that file in my branch, https://github.com/rajasekarm/babel/blob/master/packages/babel-cli/test/fixtures/babel/--copy-files%20--include-dotfiles%20with%20ignore/out-files/lib/foo/.foo.js

I dono why github shows this as removed.

Comment (nicolo-ribaudo) [reply packages/babel-cli/test/fixtures/babel/--copy-files --include-dotfiles with ignore/out-files/lib/foo/.foo.js]: The linked one is your master branch, not the PR branch :stuck_out_tongue:

File: packages/babel-cli/test/fixtures/babel/--copy-files with ignore with includeIgnored/options.json (hunk 2 above)
Comment (JLHwung) [reply packages/babel-cli/test/fixtures/babel/--copy-files with ignore with includeIgnored/options.json]: Good to have.😀

File: packages/babel-cli/test/fixtures/babel/--copy-files --include-dotfiles with ignore/out-files/lib/foo/.foo.js (hunk 3 above)
Comment (rajasekarm) [reply packages/babel-cli/test/fixtures/babel/--copy-files --include-dotfiles with ignore/out-files/lib/foo/.foo.js]: Yeah, I'm check this.

Comment (rajasekarm) [reply packages/babel-cli/test/fixtures/babel/--copy-files --include-dotfiles with ignore/out-files/lib/foo/.foo.js]: Found the root cause of this bug, I'm fixing it.

Comment (rajasekarm) [reply packages/babel-cli/test/fixtures/babel/--copy-files --include-dotfiles with ignore/out-files/lib/foo/.foo.js]: .foo.js is inside src/foo and it is in ignored list. So it is not copied.

File: packages/babel-cli/test/fixtures/babel/--copy-files with ignore with includeIgnored/options.json (hunk 2 above)
Comment (rajasekarm) [reply packages/babel-cli/test/fixtures/babel/--copy-files with ignore with includeIgnored/options.json]: Added a test case for this.

**Comment by rajasekarm** [comment]
I've exposed pathToPattern function from @babel/core for the ignore file check.

**Comment by nicolo-ribaudo** [comment]
> I've exposed pathToPattern function from @babel/core for the ignore file check.

I have a bed news for you :stuck_out_tongue: 
This is a breaking change because if someone used `@babel/core` v7.7.7 and updates `@babel/cli` to 7.8.0 won't work.
For now, we can duplicate the code (with comments to explain that it's duplicated and if it is updated it should be done in both places) and then revisit it for Babel 8.

**Comment by rajasekarm** [comment]
Sure, I'll make the change.

**Comment by rajasekarm** [comment]
@nicolo-ribaudo Done 👍

**Comment by nicolo-ribaudo** [comment]
I was looking at the `write` function in `packages/babel-cli/src/babel/dir.js`. From what I understood, it can return `false` in three cases:
1. The extension is not compatible (first `return false`)
2. The file has been ignored, and thus `babel.transformSync` (aka `util.compile`) returns `null`
3. There was an error.

If we make the `write` function return a string enum rather than a boolean, maybe we could avoid checking if the file has been ignored? Also, I think that the current approach doesn't work well if a file has been ignored not using a `cli` option, but using a config file.

**Comment by rajasekarm** [comment]
@nicolo-ribaudo I made the requested change. Can you please check.

File: packages/babel-cli/src/babel/dir.js [hunk 4]
```diff
@@ -9,6 +9,13 @@ import fs from "fs";
 import * as util from "./util";
 import { type CmdOptions } from "./options";
 
+const FILE_TYPE = Object.freeze({
+  NON_COMPACTIBLE: "NON_COMPACTIBLE",
```
Comment (JLHwung) [inline packages/babel-cli/src/babel/dir.js]: It should be `NON_COMPILABLE`.

File: packages/babel-cli/src/babel/dir.js [hunk 5]
```diff
@@ -20,11 +27,11 @@ export default async function({
 }: CmdOptions): Promise<void> {
   const filenames = cliOptions.filenames;
 
-  async function write(src: string, base: string): Promise<boolean> {
+  async function write(src: string, base: string): Promise<string> {
```
Comment (JLHwung) [inline packages/babel-cli/src/babel/dir.js]:
```suggestion
  async function write(src: string, base: string): Promise<$Keys<typeof FILE_TYPE>> {
```

File: packages/babel-cli/src/babel/dir.js (hunk 4 above)
Comment (rajasekarm) [reply packages/babel-cli/src/babel/dir.js]: Done 👍

File: packages/babel-cli/src/babel/dir.js (hunk 5 above)
Comment (rajasekarm) [reply packages/babel-cli/src/babel/dir.js]: Done 👍

**Comment by JLHwung** [comment]
@rajasekarm Can you rebase on upstream master? 🤔It should not have legacy CircleCI test check here.

**Comment by nicolo-ribaudo** [comment]
@rajasekarm Could you open a PR to babel/website for the docs?

**Comment by rajasekarm** [comment]
@nicolo-ribaudo I'll do it tonight.
